src/librustc_typeck/check/method/suggest.rs

```diff
@@ -69,13 +69,27 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
             }
 
             if !static_sources.is_empty() {
-                fcx.tcx().sess.fileline_note(
+                cx.sess.fileline_note(
                     span,
                     "found defined static methods, maybe a `self` is missing?");
 
                 report_candidates(fcx, span, item_name, static_sources);
             }
 
+            if !predicates.is_empty() {
+                let bound_list = predicates.iter()
+                    .map(|p| format!("`{} : {}`",
+                                     p.substs.self_ty().unwrap(),
```

Comment:
this unwrap is ok, but it's better still to call `p.self_ty()`, so that the unwrap is not required.

the reason it is ok is that all trait-ref substitutions should always have a self-type. Calling `self_ty` on the `TraitRef` leverages this invariant, indicating that if a failure occurs, it's because the trait-ref was malformed, not because your code is at fault.